corehq/util/es/interface.py

```diff
@@ -10,17 +10,29 @@ def __init__(self, es):
     def update_index_settings(self, index, settings_dict):
         return self.es.indices.put_settings(settings_dict, index=index)
 
+    def get_doc(self, index, doc_type, doc_id):
+        doc = self.es.get_source(index, doc_type, doc_id)
+        doc['_id'] = doc_id
+        return doc
+
+    def get_bulk_docs(self, index, doc_type, doc_ids):
+        docs = []
+        results = self.es.mget(
+            index=index, doc_type=doc_type, body={'ids': doc_ids}, _source=True)
+        for doc_result in results['docs']:
+            doc_id = doc_result['_id']
+            if doc_result['found']:
+                doc = doc_result['_source']
+                doc['_id'] = doc_id
+                docs.append(doc)
+        return docs
```

Comment:
Yeah I thought of that but wasn't sure there was much benefit to it. I guess it reduces the number of pointers we keep in memory to the same object, but I think of the biggest benefit of iterators as when they prevent us from loading all objects into memory in the first place. In this case, these objects are already all in memory from parsing the http request

Reply:
Agreed - I doubt there's an appreciable performance difference.  I like iterators because it ever-so-slightly simplifies the code in my opinion - rather than looking where `docs` comes from, where it's modified, and where it's returned, you can just see a single `yield` statement.  Pretty minor point though.

Generators are nice for data accessors too, since if this were later switched to say a chunked query, the calling code would already be equipped to handle it.  I also like to think of these things as a single stream of data with various transformations mapped over it, rather than building a series of lists.

It's 100% a style thing, and I don't think it's necessarily better or worse, was mostly flagging in case you also had a predisposition towards working with generators and just didn't notice the opportunity.

Reply:
Gotcha. I like the argument of using `yield` to return the more restrictive type (an iterator) to get other code to only ever use that more limited interface, allowing us to change the implementation to a more memory efficient one in the future.